sdk/python/feast/pyspark/launchers/k8s/k8s.py

```diff
@@ -0,0 +1,341 @@
+    ):
+        """
+        This is the job object representing the historical retrieval job, returned by KubernetesClusterLauncher.
+
+        Args:
+            output_file_uri (str): Uri to the historical feature retrieval job output file.
+        """
+        super().__init__(api, namespace, job_id)
+        self._output_file_uri = output_file_uri
+
+    def get_output_file_uri(self, timeout_sec=None, block=True):
+        if not block:
+            return self._output_file_uri
+
+        if self._wait_for_complete(timeout_sec):
+            return self._output_file_uri
+        else:
+            raise SparkJobFailure("Spark job failed")
+
+
+class KubernetesBatchIngestionJob(KubernetesJobMixin, BatchIngestionJob):
+    """
+    Ingestion job result for a k8s cluster
+    """
+
+    def __init__(self, api: CustomObjectsApi, namespace: str, job_id: str):
+        super().__init__(api, namespace, job_id)
+
+
+class KubernetesStreamIngestionJob(KubernetesJobMixin, StreamIngestionJob):
+    """
+    Ingestion streaming job for a k8s cluster
+    """
+
+    def __init__(
+        self, api: CustomObjectsApi, namespace: str, job_id: str, job_hash: str
+    ):
+        super().__init__(api, namespace, job_id)
+        self._job_hash = job_hash
+
+    def get_hash(self) -> str:
+        return self._job_hash
+
+
+class KubernetesJobLauncher(JobLauncher):
+    """
+    Submits spark jobs to a spark cluster. Currently supports only historical feature retrieval jobs.
+    """
+
+    def __init__(
+        self,
+        namespace: str,
+        incluster: bool,
+        staging_location: str,
+        resource_template_path: Optional[Path],
+    ):
+        self._namespace = namespace
+        self._api = _get_api(incluster=incluster)
+        self._staging_location = staging_location
+        if resource_template_path is not None:
+            self._resource_template = _load_resource_template(resource_template_path)
+        else:
+            self._resource_template = yaml.safe_load(DEFAULT_JOB_TEMPLATE)
+
+    def _job_from_job_info(self, job_info: JobInfo) -> SparkJob:
+        if job_info.job_type == HISTORICAL_RETRIEVAL_JOB_TYPE:
+            assert METADATA_OUTPUT_URI in job_info.extra_metadata
+            return KubernetesRetrievalJob(
+                api=self._api,
+                namespace=job_info.namespace,
+                job_id=job_info.job_id,
+                output_file_uri=job_info.extra_metadata[METADATA_OUTPUT_URI],
+            )
+        elif job_info.job_type == OFFLINE_TO_ONLINE_JOB_TYPE:
+            return KubernetesBatchIngestionJob(
+                api=self._api, namespace=job_info.namespace, job_id=job_info.job_id,
+            )
+        elif job_info.job_type == STREAM_TO_ONLINE_JOB_TYPE:
+            # job_hash must not be None for stream ingestion jobs
+            assert METADATA_JOBHASH in job_info.extra_metadata is not None
```

Comment:
By the way, are we fine with these asserts disappearing when Python code optimization is enabled? I like the conciseness of the asserts, but I am not sure if they are safe to use.

Reply:
In this case it is mostly to make mypy happy, as the code would crash on the next line anyway even without this check.